generate/avro.go

```diff
@@ -0,0 +1,304 @@
+func writeAvroField(cfg avroConfig, f *ast.Field) {
+	if len(f.Names) == 0 { // An embedded struct
+		t, ok := f.Type.(*ast.Ident)
+		if !ok {
+			return
+		}
+		newcfg := cfg
+		newcfg.namespace = t.Name
+		writeEmbeddedStructFields(newcfg)
+		return
+	}
+	name := f.Names[0].Name
+
+	tagName, tagDescription, omitEmpty := parseStructTag(f.Tag)
+
+	if tagName != "" {
+		name = tagName
+	}
+
+	avroType := fmt.Sprintf("%s", convertToAvroType(cfg, f.Type, name, omitEmpty))
+
+	avroName := fmt.Sprintf(`"name":%q`, name)
+	if cfg.namespace != "" {
+		avroName += fmt.Sprintf(`,"namespace":%q`, cfg.namespace)
+	}
+	if tagDescription != "" {
+		avroName += fmt.Sprintf(`,"doc":%q`, tagDescription)
+	}
+
+	fmt.Fprintf(cfg.writer, "{%s,%s}", avroName, avroType)
+}
+
+// writeAvroFields loops through a list of fields in a struct writing out the Avro values for the fields and adding
+// trailing commas as appropriate for JSON.
+func writeAvroFields(cfg avroConfig, list *ast.FieldList) {
+	len := list.NumFields()
+	for i, f := range list.List {
+		writeAvroField(cfg, f)
+		if i+1 != len {
+			fmt.Fprint(cfg.writer, ",")
+		}
+	}
+}
+
+// writeAvroFieldPost writes the closing characters for the various avro types as they are traversed as part of an
+// astuitl.Apply call
+func writeAvroFieldPost(w io.Writer) astutil.ApplyFunc {
+	return func(c *astutil.Cursor) bool {
+		var closer string
+		switch c.Node().(type) {
+		case *ast.TypeSpec:
+			closer = "]}"
+		}
+
+		if _, err := fmt.Fprint(w, closer); err != nil {
+			return false
+		}
+		return true
+	}
+}
+
+func writeEmbeddedStructFields(cfg avroConfig) {
+	spec, err := parseGoStruct(cfg.namespace, filepath.Join(cfg.dir, strings.ToLower(cfg.namespace)+".go"))
+	if err != nil {
+		fmt.Fprint(cfg.writer, `{"type":"embedded struct not found"}`)
+		return
+	}
+
+	astutil.Apply(spec, func(c *astutil.Cursor) bool {
+		n := c.Node()
+		if list, ok := n.(*ast.FieldList); ok {
+			writeAvroFields(cfg, list)
+			return false
+		}
+		return true
+	}, nil)
+}
+
+// convertToAvroType returns the avro type definition for a go type
+func convertToAvroType(cfg avroConfig, expr ast.Expr, name string, nullable bool) string {
+	// Note: the go code generated from JSON schema does not include maps and they are not handled here
+	switch t := expr.(type) {
+	case *ast.Ident:
+		var typeName string
+		switch t.Name {
+		case "bool":
+			typeName = "boolean"
+		case "int", "uint", "int64", "uint64":
+			typeName = "long"
+		case "int8", "int16", "int32", "uint8", "uint16", "uint32":
+			typeName = "int"
+		case "float32":
+			typeName = "float"
+		case "float64":
+			typeName = "double"
+		case "byte":
+			typeName = "bytes"
+		case "string":
+			typeName = "string"
+		default:
+			typeName = "unknown"
+		}
+		if nullable {
+			if typeName == "boolean" {
+				return fmt.Sprintf(`"type":%q,"default":false`, typeName)
+			}
+			return fmt.Sprintf(`"type":["null",%q]`, typeName)
```

Comment:
So one thing to consider, our data doesn't line up with this and will cause a serialization error when we try to convert the `map[string]interface{}` to Avro data. Perhaps this isn't an issue with the avro-gen library but the Linkedin one I was using previously was complaining. In order to work around this, I set every nullable field as not nullable with a default value for the type.

Reply:
Interesting I was using the Google documentation and the Avro schema documentation as guide and both recommend this as the approach for nullable fields.
https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-avro
http://avro.apache.org/docs/current/spec.html

Reply:
You're correct in that it's the recommended approach the problem is that our data doesn't fit this format. Rather than transforming our data to fit the format it's easier to specify the fields as non-nullable in the schema and any missing fields will have their defaults applied